### PR TITLE
Make full refresh through OvirSDK use asynchronous calls to oVirt

### PR DESCRIPTION
**Pull request by borod108**
Make full refresh through OvirSDK use asynchronous calls to oVirt to improve performance.

This should make the refresh significantly faster
https://bugzilla.redhat.com/show_bug.cgi?id=1404346

### CODE REVIEW

**Comment by borod108** [comment]
@masayag @jhernand I have applied the fixes suggested - split in to two PRs, added spec for futures_collector and made sure it will break and wait on all pending futures in case of an exception.
Please review.

**Comment by borod108** [comment]
@masayag @jhernand please review,

File: app/models/manageiq/providers/redhat/infra_manager/inventory/strategies/v4.rb [hunk 1]
```diff
@@ -42,21 +42,94 @@ def get_uuid_from_target(object)
...
+    def wait_on_top_level_collections(top_level_fututres_hash)
+      top_level_fututres_hash.each do |k, v|
+        top_level_fututres_hash[k] = v.wait
+      end
+      top_level_fututres_hash
+    end
+
+    def collect_top_level_futures
+      res = {}
+      %w(cluster storage host vm template network datacenter disk).each { |inv_name| res[inv_name.to_sym] = collect_inv_future(inv_name.to_sym) }
```
Comment (masayag) [inline app/models/manageiq/providers/redhat/infra_manager/inventory/strategies/v4.rb]: pls extract `%w(cluster storage host vm template network datacenter disk)` to constant

File: app/models/manageiq/providers/redhat/infra_manager/inventory/strategies/v4.rb [hunk 2]
```diff
@@ -107,12 +191,39 @@ def collect_vm_by_uuid(uuid)
...
     def collect_templates
       connection.system_service.templates_service.list.collect do |template|
         TemplatePreloadedAttributesDecorator.new(template, connection, preloaded_disks)
       end
     end
 
+    def collect_future_template_disk_attachments(templates, templates_service)
+      templates_attachments = templates.collect do |template|
+        template_service = templates_service.template_service(template.id)
+        atts_service = template_service.disk_attachments_service
```
Comment (masayag) [inline app/models/manageiq/providers/redhat/infra_manager/inventory/strategies/v4.rb]: `atts_service` isn't descriptive enough, can we name it `disk_attachments_service` and in correlation `disk_attachments_future`

File: app/models/manageiq/providers/redhat/infra_manager/futures_collector.rb [hunk 3]
```diff
@@ -0,0 +1,96 @@
+class ManageIQ::Providers::Redhat::InfraManager::FuturesCollector
+  attr_reader :keyed_futures_queue, :keyed_requests_queue, :parallel_processing_capacity
+  attr_accessor :res_hash
```
[the remark this answers is not in the document]
Comment (borod108) [reply app/models/manageiq/providers/redhat/infra_manager/futures_collector.rb]: done

File: app/models/manageiq/providers/redhat/infra_manager/futures_collector.rb [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    @keyed_requests_queue = []
+    @parallel_processing_capacity = args[:batch_size] || DEFAULT_PARALLEL_PROCESSING_CAPACITY
+    @res_hash = {}
+  end
+
+  def queue_keyed_request_tasks(keyed_request_tasks)
+    @keyed_requests_queue += keyed_request_tasks.collect { |kv| KeyedValue.from_hash(kv) }
+  end
+
+  def process_queues
```
[the remark this answers is not in the document]
Comment (borod108) [reply app/models/manageiq/providers/redhat/infra_manager/futures_collector.rb]: done

File: app/models/manageiq/providers/redhat/infra_manager/futures_collector.rb [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    while keyed_futures_queue.present?
+      begin
+        keyed_future = keyed_futures_queue.shift
+        res_hash[keyed_future.key] = keyed_future.value.wait
+        process_one_keyed_request if keyed_requests_queue.present?
+      end
+    end
+  end
+
+  def wait_on_all_futures
```
[the remark this answers is not in the document]
Comment (borod108) [reply app/models/manageiq/providers/redhat/infra_manager/futures_collector.rb]: done

File: app/models/manageiq/providers/redhat/infra_manager/inventory/strategies/v4.rb [hunk 6]
```diff
@@ -42,21 +42,94 @@ def get_uuid_from_target(object)
       get_uuid_from_href(object.ems_ref)
     end
 
+    ATTRIBUTES_TO_FETCH_FOR_INV_TYPE = {
```
[the remark this answers is not in the document]
Comment (borod108) [reply app/models/manageiq/providers/redhat/infra_manager/inventory/strategies/v4.rb]: done

File: app/models/manageiq/providers/redhat/infra_manager/inventory/strategies/v4.rb [hunk 7]
```diff
@@ -42,21 +42,94 @@ def get_uuid_from_target(object)
...
         @connection = connection
-        res = {}
-        res[:cluster] = collect_clusters
-        res[:storage] = collect_storages
-        res[:host] = collect_hosts
-        res[:vm] = collect_vms
-        res[:template] = collect_templates
-        res[:network] = collect_networks
-        res[:datacenter] = collect_datacenters
+        top_level_fututres_hash = collect_top_level_futures
```
[the remark this answers is not in the document]
Comment (borod108) [reply app/models/manageiq/providers/redhat/infra_manager/inventory/strategies/v4.rb]: done

File: app/models/manageiq/providers/redhat/infra_manager/inventory/strategies/v4.rb [hunk 8]
```diff
@@ -42,21 +42,94 @@ def get_uuid_from_target(object)
...
+
+    def collect_inv_with_attributes_async(inv, inv_type)
+      return inv if ATTRIBUTES_TO_FETCH_FOR_INV_TYPE[inv_type].blank?
+      inv_keyed_requests = collect_inv_attributes_request(inv, inv_type)
+      inv_fetched_attributes = ManageIQ::Providers::Redhat::InfraManager::FuturesCollector.process_keyed_requests_queue(inv_keyed_requests)
+      raise "Failed to fetch attributes for #{inv_type}" unless inv_fetched_attributes
+      set_inv_attributes!(inv, inv_fetched_attributes, inv_type)
+    end
+
+    def apply_hosts_fetched_results!(hosts, fetched_host_results)
```
[the remark this answers is not in the document]
Comment (borod108) [reply app/models/manageiq/providers/redhat/infra_manager/inventory/strategies/v4.rb]: removed it

File: app/models/manageiq/providers/redhat/infra_manager/inventory/strategies/v4.rb (hunk 2 above)
Comment (borod108) [reply app/models/manageiq/providers/redhat/infra_manager/inventory/strategies/v4.rb]: done, though I think its ok to have short names in small scopes

File: app/models/manageiq/providers/redhat/infra_manager/inventory/strategies/v4.rb (hunk 1 above)
Comment (borod108) [reply app/models/manageiq/providers/redhat/infra_manager/inventory/strategies/v4.rb]: done

**Pull request merged**